lib/AST/RequirementMachine/RewriteSystem.h

```diff
@@ -225,6 +239,12 @@ class RewriteSystem final {
   /// top-level generic signature and this array is empty.
   ArrayRef<const ProtocolDecl *> Protos;
 
+  /// The requirements written in source code.
+  std::vector<StructuralRequirement> WrittenRequirements;
```

Comment:
In a similar vein to the Errors vector, I'm wondering if its better to avoid storing this in the rewrite system, and instead pass it down from the request evaluator request to getRedundantRequirements() since it's only used there to get the Requirement and SourceLoc for the diagnostic.